pkg/agent/runner.go

```diff
@@ -1300,29 +1304,26 @@ func (s *atomicUpdateState) desiredVMState(allowDecrease bool) api.Resources {
 	//    is low so we should just decrease *anyways*.
 	//
 	// ---
-	// Now, it's worth noting that we only *barely* handle the edge cases above. We don't do it
-	// well, but we do handle them in a protocol-compliant way, and that's what counts! Eventually,
-	// this function will be rewritten, this note removed, and all will be well.
-
-	lowerBoundCU, upperBoundCU := s.computeUnitsBounds()
-
-	currentCU := upperBoundCU
-
-	// if we don't have an even compute unit *and* we're allowed to decrease, pick the middle.
-	if lowerBoundCU != upperBoundCU && allowDecrease {
-		currentCU = lowerBoundCU + (upperBoundCU-lowerBoundCU+1)/2 // +1 so we round up
-	}
+	//
+	// Broadly, the implementation works like this:
+	// 1. Based on load average, calculate the "goal" number of CPUs (and therefore compute units)
+	// 2. Cap the goal CU by min/max, etc
+	// 3. that's it!
 
-	goalCU := currentCU
-	if s.metrics.LoadAverage1Min > 0.9*float32(s.vm.Cpu.Use) {
-		goalCU *= 2
-	} else if s.metrics.LoadAverage1Min < 0.4*float32(s.vm.Cpu.Use) && allowDecrease {
-		goalCU /= 2
-	}
+	// Goal compute unit is at the point where (CPUs) × (LoadAverageFractionTarget) == (load
+	// average),
+	// which we can get by dividing LA by LAFT.
+	goalCU := uint16(math.Round(float64(s.metrics.LoadAverage1Min) / s.config.LoadAverageFractionTarget))
```

Comment:
it might be instructive/demonstrative to factor this (and some surrounding code) out into functions that you can write unittests: just so the inputs/expected outputs can be written down somewhere and validated.

Reply:
Yes, 100%. I definitely want tests for this. Currently I feel like the "agent is 5 packages stacked on top of each other pretending to be 1 package" situation is making this harder, but we can _probably_ add some tests anyways.

This will conflict with #172, so I'd like to wait until after that's done to do so. Opened #199 to track this post-merge.